synapse/axon.py

```diff
@@ -1,801 +1,1195 @@
+    '''
+    An active object that writes to disk on behalf of a blobstor (plus the client methods to interact with it)
+    '''
 
-    def metrics(self, offs=0):
+    class Command(enum.Enum):
         '''
-        Get metrics for the BlobStor. These can be aggregated to compute the storage stats.
-
-        Args:
-            offs (int): Offset to start collecting stats from.
-
-        Yields:
-            ((int, dict)): Yields index, sample data from the metrics sequence.
+        The types of commands a blobstorwriter client can issue
         '''
-        with self.lenv.begin() as xact:
-            for item in self._blob_metrics.iter(xact, offs):
-                yield item
+        WRITE_BLOB = enum.auto()
+        FINISH = enum.auto()
+        UPDATE_OFFSET = enum.auto()
 
-    def getCloneOffs(self):
+    class ClientInfo:
         '''
-        Get the current offset for the clone:index of the BlobStor.
-
-        Returns:
-            int: The offset value.
+        The session information for a single active client to a blobstor
         '''
-        with self.lenv.begin() as xact:
-
-            lval = xact.get(b'clone:indx', db=self._blob_info)
-            if lval is None:
-                return 0
-
-            return struct.unpack('>Q', lval)[0] + 1
-
-class BlobCell(s_cell.Cell):
-
-    confdefs = (
-        ('blob:mapsize', {'type': 'int', 'defval': s_const.tebibyte * 10,
-            'doc': 'The maximum size of the LMDB memory map'}),
-
-        ('blob:cloneof', {'defval': None,
-            'doc': 'The name of a blob cell to clone from'}),
-    )
-
-    def postCell(self):
-
-        if self.neuraddr is None:
-            raise s_common.BadConfValu(mesg='BlobCell requires a neuron')
-
-        path = self.getCellDir('blobs.lmdb')
-        mapsize = self.getConfOpt('blob:mapsize')
-
-        self.blobs = BlobStor(path, mapsize=mapsize)
-        self.cloneof = self.getConfOpt('blob:cloneof')
-
-        if self.cloneof is not None:
-            self.cellpool.add(self.cloneof, self._fireCloneThread)
-            self.cellinfo['blob:cloneof'] = self.cloneof
-
-    def finiCell(self):
-        self.blobs.fini()
-
-    def _saveDispItems(self, items):
-
-        def genr():
-
-            for link, mesg in items:
+        BUF_SIZE = 16 * s_const.mebibyte
+
+        def __init__(self):
+            self.starttime = s_common.now()
+            self.newhashes = []
+            self.totalsize = 0
+            self.nextfile()
+
+        def nextfile(self):
+            self.nextchunknum = 0
+            self.tmpfh = tempfile.SpooledTemporaryFile(buffering=self.BUF_SIZE)
+            self.hashing = hashlib.sha256()
+
+    def __init__(self, blobstor):
+        s_coro.Fini.__init__(self)
+        self.xact = IncrementalTransaction(blobstor.lenv)
+        self.lenv = blobstor.lenv
+        self.blobstor = blobstor
+        self._workq = _AsyncQueue(50)
+        self._worker = threading.Thread(target=self._workloop, name='BlobStorWriter')
```

Comment:
another case where we could use @firethread decorator

Reply:
done